---
title: "comaprExamples"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{comaprExamples}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
author: Ruqian Lyu
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 8,
  fig.height = 8
)
```

```{r setup}
devtools::load_all()

suppressPackageStartupMessages({
  library(readxl)
  library(ggplot2)
  library(dplyr)
  library(reshape2)
  library(chromoMap)
  library(UpSetR)
  library(ggrepel)
  library(comapr)
  library(ggplot2)
  library(reshape2)
  library(dplyr)
})
```


## Load data sheet

We load the SNP genotyping data from excel spreadsheet, here we are using the sheet "mutant_25-6-19_informative" from All_data_May_to_August_2019.xlsx.

```{r}

mutate_inform1 <- read_excel(path = "/mnt/mcfiles/rlyu/Projects/Snakemake_projects/yeln_2019_spermtyping/analysis/All_data_May_to_August_2019.xlsx",
                             sheet = "mutant_25-6-19_informative")
mutate_inform1 <- data.frame(mutate_inform1)

```

Check if we have any duplicated markers names by checking any duplicate `CHR_POS`:

Then we name the markers using the combination of CHR and POS.
```{r}
mutate_inform1[duplicated(paste0(mutate_inform1$CHR,"_",mutate_inform1$POS)),]
mutate_inform1$CHR_POS <- paste0(mutate_inform1$CHR,"_",mutate_inform1$POS)
```


## Example for mutate_inform1

```{r}
rownames(mutate_inform1) <- mutate_inform1$CHR_POS

```

### Get the genotype matrix

The genotype matrix is a matrix of markers (rows) by samples(columns): 

```{r}
gt_matrix <- mutate_inform1[,grep("[0-9].*[0-9]$",colnames(mutate_inform1))]

head(gt_matrix)
```

`plotMissingGT` generates dot plots plot of markers by samples indicating which markers are missing from which samples.

```{r}
plotMissingGT(gt_matrix = gt_matrix, plot_wg = FALSE)+theme_classic()+ 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

### Correct the geneotypes 

Next, we correct the geneotypes which includes 1), change genotype to labels, 2) infer the missing genotypes, 3) change "Home_ref" to "Het".

```{r}
#rownames(gt_matrix)
gt_matrix <- correctGT(gt_matrix = gt_matrix,
                       ref = mutate_inform1$C57BL.6J,
                       alt = mutate_inform1$FVB.NJ..i.)
```

Then we plot the missing genotype again, there are still some SNP markers that have missing values:

```{r}
plotMissingGT(gt_matrix = gt_matrix, plot_wg = FALSE)+theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

### Filter genotype matrix 

We need to remove the missing markers to make the entire dataset suitable for statistical testing (i.e for testing difference in genetic lengths in contrasting groups).

We can either remove markers with missing genotype or samples with any missing genotype. 

Let's first check how many markers have missing values/ how many samples have missing values:

```{r}
cts <- countGT(gt_matrix,plot = TRUE, interactive = FALSE)
```

We can also generate interactive plot:

```{r}
ply_cts <- countGT(gt_matrix,plot = TRUE, interactive = TRUE)
ply_cts$ply
```

Filter samples:

```{r}


#p$plot
gt_matrix_filtered <- filterGT(gt_matrix,min_markers = 245)

p <- countGT(gt_matrix_filtered,plot = TRUE, interactive = FALSE)
```

Filter markers

```{r}
#p$plot
## filter NA markers
## The idea here is because we will do K-S test on the kosambi genetic lengths calculated we would like to make sure they are derived on the
## same set of marker intervals


gt_matrix_filtered <- filterGT(gt_matrix,min_samples = 14)

p <- countGT(gt_matrix_filtered,plot = TRUE, interactive = FALSE)
```


We choose to retain all samples and filter out some markers. We now have 251 markers by 22 samples

Check there is no NA in the genotype matrix anymore:

We still have some NA genotypes in the genotype matrix.
```{r}
sum(apply(gt_matrix_filtered,1,anyNA))
```

### Find duplicated samples 

We next check if any samples are duplicated by checking the pair-wise sample genotype similarity:

```{r}

## Find duplicated samples
dups_samples <- findDupSamples(gt_matrix_filtered)
## 
```

Sample X99 and X98 are duplicated, remove one of them

```{r}
gt_matrix_filtered <- gt_matrix_filtered[,colnames(gt_matrix_filtered) !="X98"]

```

Check there is no duplicated sample anymore:

```{r}
dup_samples <- findDupSamples(gt_matrix_filtered)
dup_samples
```

### Find duplicated markers

We have some markers that almost always have the same genotype across samples but not exactly the same.

We are not removing duplicated markers for now.

```{r,fig.width=10}

dup_markers <- findDupMarkers(gt_matrix_filtered,threshold = 1)
dup_markers
## Detect crossover by examing the genotype pattern
```
### Find distorted markers

Marker disortation detection using chisq-test
We expect the genotypes to appear with the frequenceis of 1:1 homo_alt:hets, we
use chisq-test for finding markers that have genotypes significantly diffferent from
the 1:1 ratio and report them


The significantly distorted markers are markers on X chromosome:

```{r}
distor_markers <- getDistortedMarkers(gt_matrix_filtered)
distor_markers[distor_markers$Adj.pvals < 0.05,]
```

### Detect crossovers

We find crossovers within each marker intervals across all samples by calling the `detectCO` function. 

For intervals tha have NA values for some samples, they are due to the genotypes for relevant SNP markers are missing.

```{r}
gt_matrix_co <- detectCO(gt_matrix = gt_matrix_filtered,
                          chrs = sapply(strsplit(rownames(gt_matrix_filtered),"_"),`[[`,1),
                          chrPos = sapply(strsplit(rownames(gt_matrix_filtered),"_"),`[[`,2),
                          type = "bool")

head(gt_matrix_co)
```

If we set the type of the returned matrix as "counts", this function returns a matrix with cumulative crossover counts.

```{r}
gt_matrix_co_counts <- detectCO(gt_matrix = gt_matrix_filtered,
                          chrs = sapply(strsplit(rownames(gt_matrix_filtered),"_"),`[[`,1),
                          chrPos = sapply(strsplit(rownames(gt_matrix_filtered),"_"),`[[`,2),
                          type = "counts")

head(gt_matrix_co_counts)

```

The row names of this `gt_matrix_co` indicating the marker interval ID which tells "Chr_Start_End" of the interval. 
"firstM" indicates the interval is the fisrt SNP marker on this chromosome.

Next we prepare the matrix for feeding into `calGeneticMap` function as below:

```{r}
gt_matrix_co$interval_ID <- rownames(gt_matrix_co)
gt_matrix_co_by_marker <- melt(gt_matrix_co, id = "interval_ID")

colnames(gt_matrix_co_by_marker) <- c("interval_ID","Sample","Cross_over")

gt_matrix_co_by_marker$interval_ID <- as.character(gt_matrix_co_by_marker$interval_ID)

head(gt_matrix_co_by_marker)
```

### Plot Genotype Frequency

We expect the samples' frequecies of hets and homo_alternative genotypes should be around 0.5.

```{r}
plotGTFreq(gt_matrix_filtered)+facet_wrap(.~variable)+theme(axis.text.x = element_text(
  angle = 90
))
```

### Calculate genetic map

Then we call `calGeneticMap` which generates the point estimates of crossover rate, upper_ci, lower_ci and the mapped kosambi/haldane genetic lengths for each interval.

```{r}
#head(gt_matrix_co_by_marker)

## Problems in getting kosambi confidence interval, NaN produced

gt_matrix_dist <- calGeneticMap(gt_matrix_co_by_marker = gt_matrix_co_by_marker )

head(gt_matrix_dist)
```

We now order the intervals by chromosomes then by starting positions of the interval.

```{r}
gt_matrix_dist$CHR  <-  sapply(strsplit(gt_matrix_dist$interval_ID,"_"),`[[`,1)

gt_matrix_dist$POS <-   as.numeric(sapply(strsplit(gt_matrix_dist$interval_ID,"_"),`[[`,3))

gt_matrix_dist$CHR  <-  factor(gt_matrix_dist$CHR,levels = c(seq(1:19),"X"))

## important to order the markers by chr then by position
gt_matrix_dist <- gt_matrix_dist[order(gt_matrix_dist$CHR, rank(gt_matrix_dist$POS,na.last = FALSE)), ]
```

We derive the cumulative haldane and kosambi genetic lengths per chromosome:

```{r}
gt_matrix_final <- gt_matrix_dist %>% group_by(CHR) %>% mutate(cum_haldane = cumsum(haldane),
                                                              cum_kosambi = cumsum(kosambi))

#gt_matrix_dst[sort(gt_matrix_dst$CHR,gt_matrix_dst$POS),]
head(gt_matrix_final)

```

total centimorgans by kosambi for mutant_25-6-19_informative

```{r}
k_sum  <- sum(gt_matrix_final$kosambi*100)
k_sum
```

### Plots

```{r}
ggplot(data = gt_matrix_final)+geom_point(mapping = aes(x = POS, y = cum_haldane ),color = "blue")+
  facet_wrap(.~CHR)+geom_point(mapping = aes(x = POS, y = cum_kosambi ),color = "red")+ggtitle("cum_haldane (blue),cum_kosambi (red)")+
  theme(axis.text.x = element_blank())+ylab(label = "cumulative genetic length")
#
#tt <- gt_matrix_final %>% group_by(CHR) %>% mutate(sum_chr = sum(kosambi))



# ggplot(data = gt_matrix_final)+geom_point(mapping = aes(x = CHR,y = cum_kosambi*100),stat = "identity")+
#   geom_label_repel(mapping = aes(x = CHR,y = cum_kosambi*100,label = interval_ID))+
#   geom_bar(data = tt[!duplicated(tt$CHR),],mapping = aes(y = CHR,x = sum_chr*100),
#            width = .5,stat = "identity",alpha  = 0.2)
```
```{r,fig.width=10}
plotdf <- gt_matrix_final
plotdf$interval_ID <- factor(plotdf$interval_ID, levels = plotdf$interval_ID[order(plotdf$pointEst)] )
ggplot(data = plotdf)+geom_point(mapping = aes(x = interval_ID, y = pointEst))+geom_errorbar(
  mapping = aes(ymax = upper_ci, ymin = lower_ci,x = interval_ID,
                colour = CHR)
)+theme_classic()+theme(axis.text.x = element_blank())+ylab("crossover rate point estimate and confidence interval")+facet_wrap(.~CHR,
                                                                                                                                scales = "free")
```

Total genetic length

```{r,fig.width=10}
plotdf <- gt_matrix_final

ggplot(data = plotdf)+geom_point(mapping = aes(x = interval_ID, y = cum_kosambi))+theme_classic()+theme(axis.text.x = element_blank())+
  ylab("kosambi genetic lengths per interval")+facet_wrap(.~CHR)
    

plot_df <- gt_matrix_final
########
head(plot_df)
plot_df <- plot_df[order(plot_df$CHR, xtfrm(plot_df$POS)), ]
head(plot_df)
ggplot(data = plot_df)+geom_point(mapping = aes(x = POS, y= cum_kosambi))+facet_wrap(.~CHR)+
  ggtitle(label = paste0("total genetic length Mutant June: \n",round(sum(plot_df$kosambi*100),2)))+
  ylab("cumulative kosambi genetic lengths per chromosome")



df.tmp <- plot_df %>%

  # Compute chromosome size
  group_by(CHR) %>%
  summarise(chr_len=max(POS)) %>%

  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%

  # Add this info to the initial dataset
  left_join(plot_df, ., by=c("CHR"="CHR")) %>%

  # Add a cumulative position of each SNP
  arrange(CHR, POS) %>%
  mutate( BPcum=POS+tot)
# get chromosome center positions for x-axis
axisdf <- df.tmp %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
df.tmp$wg_cum_kosambi <- 0

df.tmp$wg_cum_kosambi <- cumsum(df.tmp$kosambi)

ggplot(data = df.tmp)+geom_point(mapping = aes(x = BPcum, y= wg_cum_kosambi))+
  ggtitle(label = paste0("total genetic length Mutant June: \n",
                         round(sum(plot_df$kosambi)*100,2)))+
  theme(axis.text.x  = element_text(angle = 90, hjust = 1))+scale_x_continuous( label = axisdf$CHR, breaks= axisdf$center)+
  ylab("Whole Genome cumulative kosambi genetic length")+theme_classic()

```


```{r}
saveRDS(gt_matrix_final,"../data/mutant_june_gt_final.rds")
#saveRDS(gt_matrix_final,"../data/wildtype_may_gt_final.rds")

```


