---
title: "GetStarted_GeneticDsitancesCalculation"
output: rmarkdown::html_document
vignette: >
  %\VignetteIndexEntry{Get-Started-With-comapr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  
author: Ruqian Lyu
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(comapr)
#devtools::load_all()
library(GenomicRanges)
library(BiocParallel)
#library(Gviz)
```


# Introduction

`compar` lets you manipulate the genotyping results for a sequence of ordered 
markers across the genome. It integrates with modern bioconductor packages and
provide various visualisation function. In this document, we walk through the
basic functions that derive the genetic distances for a demo dataset.


# The small demo data set

The package includes a small data object that contains 70 markers and 22 samples.
The 22 samples are F1 progenies from crossing a C57BL/6J mouse and a FVB/NJ 
mouse whose genotypes are also included in this `snp_geno` data object.

```{r}
BiocParallel::register(SerialParam())
BiocParallel::bpparam()
```

```{r}
data(snp_geno_gr)
data(parents_geno)
```

### Format the genotype

In order to detect crossovers from the 22 samples' genotype results, we'd need
to format the result into `Homo_ref`, `Homo_alt` and `Het` encoding. This can be
simply done through comparing with the parents' genotypes.

```{r}
corrected_geno <- correctGT(gt_matrix = mcols(snp_geno_gr),
                            ref = parents_geno$ref,
                            alt = parents_geno$alt)
mcols(snp_geno_gr) <- corrected_geno
```

The `corrected_geno` matrix 

```{r}
head(mcols(snp_geno_gr))

```

Note that there are missing values in this resulting matrix that can be resulted
from:

- No genotype data from the genotyping result
- The genotype is wrong and it is converted to `NA`

### Find outlier samples/markers

```{r}
genotype_counts <- countGT(mcols(snp_geno_gr))
genotype_counts$plot
```
```{r}
genotype_counts$n_markers
genotype_counts$n_samples
```

Filter out markers/samples by `filterGT`.

A printed message contains information about how much markers or samples have 
been filtered.

```{r}
corrected_geno <- filterGT(snp_geno_gr,min_markers = 30,
                           min_samples = 2)
```

### Find sample duplicates

Sample duplicates are identified by finding samples that share exactly same genotypes
across all available markers.

```{r}
dups <- findDupSamples(mcols(corrected_geno),threshold = 0.99)
dups
```
Remove the duplicated samples

```{r}
mcols(corrected_geno) <- mcols(corrected_geno)[,colnames(mcols(corrected_geno))!="X98"]
#corrected_geno
```


### Count crossovers

Crossovers are detected and counted through examining the patterns of genotypes 
in each sample against the list of markers. When there is a shift from one 
genotype block to another the crossovers are detected. This is done through calling `countCOs` function.

```{r}
marker_gr_cos <- countCOs(corrected_geno)

marker_gr_cos[1:5,]
```

### Calculate genetic distance 

The genetic distances via mapping function, `Kosambi` or `Haldane` are calculated
based on the crossover rates. by calling the `calGeneticDist` function, the 
centiMorgan distances of marker intervals are returned.


```{r}
dist_gr <- calGeneticDist(marker_gr_cos)

dist_gr[1:5,]
```

###  Calculate genetic distance for binned intervals 

Alternative, instead of returning the genetic distances in marker intervals, you
can specify a `bin_size` that tells `calGeneticDist` to return calculated genetic
distances for equally sized chromosome bins.

```{r}
dist_bin_gr <- calGeneticDist(marker_gr_cos,bin_size = 1e6)

dist_bin_gr[1:5,]
```

### Plotting genetic distance (in centiMorgans) for each bin

```{r}
plotGeneticDist(dist_bin_gr,chr = "1")
```


### Plot cumulative distances

Plot the cumulative genetic distances of certain chromosomes

```{r}
plotGeneticDist(dist_bin_gr,chr="1",cumulative = T)

```

Multiple chromosomes

```{r,fig.height=10,fig.width=8}
plotGeneticDist(dist_bin_gr,cumulative = T)
```

### Whole Genome Genetic distances Plot

```{r}
plotWholeGenome(dist_bin_gr)
```

# Sessioninfo

```{r sessioninfo}
sessionInfo()
```



